ast/ast.go

```diff
@@ -184,18 +186,27 @@ type SubGraph struct {
 }
 
 func NewSubGraph(id, l Attrib) (*SubGraph, error) {
-	g := &SubGraph{ID: ID(fmt.Sprintf("anon%d", r.Int63()))}
-	if id != nil {
-		if len(id.(ID)) > 0 {
-			g.ID = id.(ID)
-		}
+	g := &SubGraph{}
+	if idTyped, ok := id.(ID); id == nil || (ok && len(idTyped) == 0) {
+		g.ID = ID(fmt.Sprintf("anon%d", randInt63()))
+	} else if ok && (len(idTyped) > 0) {
+		g.ID = idTyped
+	} else if id != nil && !ok {
```

Comment:
This can probably just be else, to make sure we covered all cases.
My brain is having a hard time following the ifs.

Reply:
Yeah it's a whole forest of branches now.